manifests/key.pp

```diff
@@ -0,0 +1,43 @@
+# Generate a new key for the dns
+#
+# === Parameters:
+#
+# $secret::                     This is the secret to be place inside the keyfile, if left empty the key will be generated
+#    
+define dns::key(
+  String               $algorithm    = 'hmac-md5',
+  String               $filename     = "${name}.key",
+  Optional[String]     $secret       = undef,
+  Stdlib::Absolutepath $keydir       = $::dns::dnsdir,
+) {
+  $keyfilename = "${keydir}/${filename}"
+
+  if ! $secret {
+    exec { "create-${filename}":
+      command => "${dns::rndcconfgen} -r /dev/urandom -a -c ${keyfilename} -b 512 -k ${name}",
+      creates => $keyfilename,
+      notify  => Service[$::dns::namedservicename],
+    }-> file { $keyfilename:
+      owner => 'root',
+      group => $dns::params::group,
+      mode  => '0640',
+    }
+  }
+
+  if $secret {
+    file {$keyfilename:
+      ensure  => file,
+      owner   => $dns::user,
+      group   => $dns::group,
+      mode    => '0640',
+      content => template('dns/key.erb'),
+      notify  => Service[$::dns::namedservicename],
+    }
+  }
+
+  concat::fragment { 'named.conf+20-keys.dns':
```

Comment:
You should include the key name in here because otherwise you get duplicate resource definitions when ```dns::key``` is used multiple times.